### ISSUE TITLE
multiple hosting sites with emulator

### ISSUE THREAD

**Issue by ZYinMD** [firebase/firebase-tools#2570]
If I configure multiple hosting sites with "deploy targets", and run the emulator, there are 2 problems:

1. The emulator UI will only show the link to the default site (the one at `https://project-id.web.app`). The other site(s) are served, but no link is shown. This should be an easy fix. (btw if you googled your way here, please know they're indeed served, good news right?)
2. Only the port for the main site is customizable in firebase.json, the other sites will get auto assigned ports (they seems to start from main site +5, then +6 +7), there's no way to customize.

**Comment by ZYinMD** [comment]
update: I just found out the emulator is indeed hosting multiple sites for me. The problem is:
1. the emulator UI only renders a link to the 1st hosting site, not the rest
2. there's no way to assign the port in firebase.json, currently the rest of the ports are auto assigned 5 digits apart from the 1st hosting site

**Comment by neekey** [comment]
any update to solve this issue yet?

**Comment by danmt** [comment]
@ZYinMD I had the same issue. I solved it for my use case with the `--only` option. I run emulators this way:

```bash
firebase emulators:start --only "hosting:<target>" --project <projectId>
```

In my case, each site belongs to a different project but I suppose you can avoid the `--project` option, and firebase defaults to the project's default in `.firebaserc`. Hope that helps.

**Comment by motiejunas** [comment]
> kmcnellis self-assigned this on 3 Oct 2020
> kmcnellis unassigned kmcnellis on 28 Dec 2020

okay... 😃

**Comment by gpfister** [comment]
Hi ! I have the same reqest, in particular to test (and include in CI workflow) cross web app integration. Many thanks in advance.

**Comment by rcb4t2** [comment]
+1 from us! We need to be able to specify an emulator port for each of the hosting targets defined in our config

In our use case, we have rewrites to triggered functions on one of our hosting targets. In order to test in development, we have to duplicate those rewrites over to the main target. Since we assign the url's programmatically, it's not feasible to fetch the port from the firebase-tools console output

**Comment by kuechlerm** [comment]
Hi. Any updates here? I would really like to have this for testing purposes.

**Comment by s2young** [comment]
+1

**Comment by charlesgreen** [comment]
Hi, Any updates? This would be helpful for testing.

**Comment by charlesgreen** [comment]
Hi @bkendall, I'm interested in contributing. Is the team open to receiving a PR for these changes?

**Comment by bkendall** [comment]
Yeah, I think we can review a PR. I'm not 100% sure what the best solution would look like off the top of my head, but I'd be happy to work on that in a PR (though, be warned, my CLI time is pretty limited these days so I'm sorry if I'm not immediately responsive)

**Comment by mvarchdev** [comment]
Hello all, Is there any update? I have an issue when I have --project demo-.... prefix with using emulators, that I am not able with any tryings to get emulators with multi site running. Any suggestions? Thanks in advance. @charlesgreen @bkendall

**Comment by wneel** [comment]
Hello, if anyone is still wondering how it's going, here’s a little summary:

> they seems to start from main site +5, then +6 +7)

Yes, here you can find why. This prevents overlapping with the Functions' reserved ports:
https://github.com/firebase/firebase-tools/blob/58c383c9b3682ad9bef6b57b5c0e4034d54aa811/src/serve/hosting.ts#L135

So you can easily access every website in your `firebase.json` by adjusting the ports in the URL.

After running `firebase emulators:start`, for each correctly configured website, you will see an output like this:
```
hosting[foo]: Local server: http://127.0.0.1:5002
```
Here, **foo** is defined by `config.site` or `config.target`, as you can find here:
https://github.com/firebase/firebase-tools/blob/58c383c9b3682ad9bef6b57b5c0e4034d54aa811/src/serve/hosting.ts#L73

---

The problem was initially that the links were not communicated and that only the port of the first site was customisable, but now that the links are communicated let's take a look at the ports

Because of the current architecture for validating ports and the type of objects expected, it would require a major refactor to customise the ports when you can just add + 5, 6, 7, etc to the basic port. It would still be nice to have the different sites served in the UI, but I think that this issue can be safely closed because the sites are now available as planned.
